### ISSUE TITLE
illegal character 0

### ISSUE THREAD

**Issue by superantichrist** [dptech-corp/Uni-Fold#15]
This is my output from docker container.

root@c305fc365212:/home/unifold# bash run_unifold.sh /home/unifold/T1052.fasta /home/unifold/output /home/data 2020-05-01 model_2_ft /home/unifold/params/monomer.unifold.pt
Starting homogeneous searching...
I0808 22:00:42.066362 139658661553856 templates.py:945] Using precomputed obsolete pdbs /home/data/pdb_mmcif/obsolete.dat.
I0808 22:00:42.070091 139658661553856 homo_search.py:160] searching homogeneous Sequences & structures for unifold...
I0808 22:00:42.070847 139658661553856 jackhmmer.py:140] Launching subprocess "/usr/bin/jackhmmer -o /dev/null -A /tmp/tmpezbgmjgz/output.sto --noali --F1 0.0005 --F2 5e-05 --F3 5e-07 --incE 0.0001 -E 0.0001 --cpu 8 -N 1 /home/unifold/T1052.fasta /home/data/uniref90/uniref90.fasta"
I0808 22:00:42.087588 139658661553856 utils.py:36] Started Jackhmmer (uniref90.fasta) query
I0808 22:06:28.933336 139658661553856 utils.py:40] Finished Jackhmmer (uniref90.fasta) query in 346.845 seconds
I0808 22:06:28.936954 139658661553856 jackhmmer.py:140] Launching subprocess "/usr/bin/jackhmmer -o /dev/null -A /tmp/tmp2phimc6z/output.sto --noali --F1 0.0005 --F2 5e-05 --F3 5e-07 --incE 0.0001 -E 0.0001 --cpu 8 -N 1 /home/unifold/T1052.fasta /home/data/mgnify/mgy_clusters_2018_12.fa"
I0808 22:06:28.955174 139658661553856 utils.py:36] Started Jackhmmer (mgy_clusters_2018_12.fa) query
I0808 22:12:37.104840 139658661553856 utils.py:40] Finished Jackhmmer (mgy_clusters_2018_12.fa) query in 368.149 seconds
I0808 22:12:37.155930 139658661553856 hmmbuild.py:121] Launching subprocess ['/usr/bin/hmmbuild', '--hand', '--amino', '/tmp/tmpvx3aub12/output.hmm', '/tmp/tmpvx3aub12/query.msa']
I0808 22:12:37.172153 139658661553856 utils.py:36] Started hmmbuild query
I0808 22:12:37.564175 139658661553856 hmmbuild.py:129] hmmbuild stdout:
# hmmbuild :: profile HMM construction from multiple sequence alignments
# HMMER 3.3 (Nov 2019); http://hmmer.org/
# Copyright (C) 2019 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# input alignment file:             /tmp/tmpvx3aub12/query.msa
# output HMM file:                  /tmp/tmpvx3aub12/output.hmm
# input alignment is asserted as:   protein
# model architecture construction:  hand-specified by RF annotation
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

# idx name                  nseq  alen  mlen eff_nseq re/pos description
#---- -------------------- ----- ----- ----- -------- ------ -----------
1     query                  198  1290   832     5.75  0.590 

# CPU time: 0.38u 0.00s 00:00:00.38 Elapsed: 00:00:00.39

stderr:

I0808 22:12:37.564431 139658661553856 utils.py:40] Finished hmmbuild query in 0.392 seconds
I0808 22:12:37.566694 139658661553856 hmmsearch.py:117] Launching sub-process ['/usr/bin/hmmsearch', '--noali', '--cpu', '8', '--F1', '0.1', '--F2', '0.1', '--F3', '0.1', '--incE', '100', '-E', '100', '--domE', '100', '--incdomE', '100', '-A', '/tmp/tmpil8uqp24/output.sto', '/tmp/tmpil8uqp24/query.hmm', '/home/data/pdb_seqres/pdb_seqres.txt']
I0808 22:12:37.579384 139658661553856 utils.py:36] Started hmmsearch (pdb_seqres.txt) query
I0808 22:12:49.915640 139658661553856 utils.py:40] Finished hmmsearch (pdb_seqres.txt) query in 12.336 seconds
Traceback (most recent call last):
  File "unifold/homo_search.py", line 306, in <module>
    app.run(main)
  File "/opt/conda/lib/python3.8/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "/opt/conda/lib/python3.8/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "unifold/homo_search.py", line 284, in main
    generate_pkl_features(
  File "unifold/homo_search.py", line 176, in generate_pkl_features
    feature_dict = data_pipeline.process(
  File "/home/unifold/unifold/msa/pipeline.py", line 193, in process
    pdb_templates_result = self.template_searcher.query(msa_for_templates)
  File "/home/unifold/unifold/msa/tools/hmmsearch.py", line 89, in query
    return self.query_with_hmm(hmm)
  File "/home/unifold/unifold/msa/tools/hmmsearch.py", line 128, in query_with_hmm
    raise RuntimeError(
RuntimeError: hmmsearch failed:
stdout:
# hmmsearch :: search profile(s) against a sequence database
# HMMER 3.3 (Nov 2019); http://hmmer.org/
# Copyright (C) 2019 Howard Hughes Medical Institute.
# Freely distributed under the BSD open source license.
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# query HMM file:                  /tmp/tmpil8uqp24/query.hmm
# target sequence database:        /home/data/pdb_seqres/pdb_seqres.txt
# MSA of all hits saved to file:   /tmp/tmpil8uqp24/output.sto
# show alignments in output:       no
# sequence reporting threshold:    E-value <= 100
# domain reporting threshold:      E-value <= 100
# sequence inclusion threshold:    E-value <= 100
# domain inclusion threshold:      E-value <= 100
# MSV filter P threshold:       <= 0.1
# Vit filter P threshold:       <= 0.1
# Fwd filter P threshold:       <= 0.1
# number of worker threads:        8
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Query:       query  [M=832]

stderr:
Parse failed (sequence file /home/data/pdb_seqres/pdb_seqres.txt):
Line 1360234: illegal character 0

Starting prediction...
start to load params /home/unifold/params/monomer.unifold.pt
start to predict unifold
Traceback (most recent call last):
  File "unifold/inference.py", line 232, in <module>
    main(args)
  File "unifold/inference.py", line 103, in main
    batch = load_feature_for_one_target(
  File "unifold/inference.py", line 46, in load_feature_for_one_target
    batch, _ = load_and_process(
  File "/home/unifold/unifold/dataset.py", line 234, in load_and_process
    features, labels = load(**load_kwargs, is_monomer=is_monomer)
  File "/home/unifold/unifold/dataset.py", line 129, in load
    all_chain_features = [
  File "/home/unifold/unifold/dataset.py", line 130, in <listcomp>
    load_single_feature(s, monomer_feature_dir, uniprot_msa_dir, is_monomer)
  File "/home/unifold/unifold/data/utils.py", line 33, in wrapper
    return copy_lib.copy(cached_func(*args, **kwargs))
  File "/home/unifold/unifold/dataset.py", line 72, in load_single_feature
    monomer_feature = utils.load_pickle(
  File "/home/unifold/unifold/data/utils.py", line 33, in wrapper
    return copy_lib.copy(cached_func(*args, **kwargs))
  File "/home/unifold/unifold/data/utils.py", line 67, in load_pickle
    ret = load(path)
  File "/home/unifold/unifold/data/utils.py", line 64, in load
    with open_fn(path, "rb") as f:
  File "/opt/conda/lib/python3.8/gzip.py", line 58, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)
  File "/opt/conda/lib/python3.8/gzip.py", line 173, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
FileNotFoundError: [Errno 2] No such file or directory: '/home/unifold/output/unifold/A.feature.pkl.gz'

**Comment by ZiyaoLi** [comment]
It seems that the parsing of file `pdb_seqres.txt` failed:

```
Parse failed (sequence file /home/data/pdb_seqres/pdb_seqres.txt):
Line 1360234: illegal character 0
```

Would you please upload the file you used? I understand that the file is big, so perhaps you can locate Line 1360234 and show the context nearby.

**Comment by ZiyaoLi** [comment]
Also @BaozCWJ may look into this.

**Comment by superantichrist** [comment]
>7ooo_B mol:na length:11  DNA (5'-D(*CP*TP*(RWQ)P*TP*CP*TP*TP*TP*G)-3')
CT05ATCTTTG
>7ooo_E mol:na length:11  DNA (5'-D(*CP*TP*(RWQ)P*TP*CP*TP*TP*TP*G)-3')
CT05ATCTTTG

second line is Line 1360234, and as error says, '0' character involved.

I used databases from alphafold2 reference site, and it works with their code.

**Comment by ZiyaoLi** [comment]
Copy that. It looks that the script mishandled the DNA chains as fastas so that 0 is invalid. We are looking into the solution.

**Comment by BaozCWJ** [comment]
> > 7ooo_B mol:na length:11  DNA (5'-D(_CP_TP*(RWQ)P_TP_CP_TP_TP_TP_G)-3')
> > CT05ATCTTTG
> > 7ooo_E mol:na length:11  DNA (5'-D(_CP_TP*(RWQ)P_TP_CP_TP_TP_TP_G)-3')
> > CT05ATCTTTG
> 
> second line is Line 1360234, and as error says, '0' character involved.
> 
> I used databases from alphafold2 reference site, and it works with their code.

It looks like the error of hmmsearch command
`/usr/bin/hmmsearch --noali --cpu 8 --F1 0.1 --F2 0.1 --F3 0.1 --incE 100 -E 100 --domE 100 --incdomE 100 -A /tmp/tmpil8uqp24/output.sto /tmp/tmpil8uqp24/query.hmm /home/data/pdb_seqres/pdb_seqres.txt`
could you please try the same fasta with alphafold's code and upload the coressponding hmmsearch command?

**Comment by superantichrist** [comment]
I checked alphafold code and It appears hmmsearch only used in multimer model, and when I run multimer, same error occured.

command seems same. (see below)
I0809 13:25:16.533803 140494127470400 run_docker.py:255] I0809 04:25:16.531341 139681435416384 hmmsearch.py:103] Launching sub-process ['/usr/bin/hmmsearch', '--noali', '--cpu', '8', '--F1', '0.1', '--F2', '0.1', '--F3', '0.1', '--incE', '100', '-E', '100', '--domE', '100', '--incdomE', '100', '-A', '/tmp/tmp69wsql5x/output.sto', '/tmp/tmp69wsql5x/query.hmm', '/mnt/pdb_seqres_database_path/pdb_seqres.txt']
....(omitted)
I0809 13:25:28.802923 140494127470400 run_docker.py:255] stderr:
I0809 13:25:28.803056 140494127470400 run_docker.py:255] Parse failed (sequence file /mnt/pdb_seqres_database_path/pdb_seqres.txt):
I0809 13:25:28.803190 140494127470400 run_docker.py:255] Line 1360234: illegal character 0

Sorry for inaccurate info that alphafold2 ref works.

So, Is my pdb_seqres.txt file a problem?

**Comment by superantichrist** [comment]
ftp://ftp.wwpdb.org/pub/pdb/derived_data/pdb_seqres.txt seems modified. I checked my old pdb_seqres.txt and find differences.
I re-run with old one and error not occured!
Thanks for your reply.

**Comment by ZiyaoLi** [comment]
Glad to see this solved, and thank you for the update. I believe they do not include dna & rna in the pdb_seqres.txt now, which is the reason of the problem.

**Comment by dominik-handler** [comment]
Hi, I ran into the same problem. Is there a fix that works with the new PDB_seqres version? Or alternatively, do you know if the old file is still available somewhere? 

Thank you,
Dominik

**Comment by dominik-handler** [comment]
I was able to fix my problem by manually removing all sequences containing invalid characters according to this: 
https://github.com/deepmind/alphafold/issues/569

tac pdb_seqres.txt | sed '/^CT05/,+1d' | tac > pdb_seqres_fixed.txt

Dominik

**Comment by ZiyaoLi** [comment]
@dominik-handler  Thank you for the report, and the hot-fix for this issue. Since it's been reported twice, we are now looking into a solution to automatically fix this issue.

**Comment by dominik-handler** [comment]
Thank you! 
By the way the database file-structure created by download_all does not fit the expected file naming of run_unifold.sh. 
I had to manually fix the structure of the uniprot database.

**Comment by ZiyaoLi** [comment]
> Thank you! By the way the database file-structure created by download_all does not fit the expected file naming of run_unifold.sh. I had to manually fix the structure of the uniprot database.

Maybe @BaozCWJ can look into this?

**Comment by BaozCWJ** [comment]
> Thank you!
> By the way the database file-structure created by download_all does not fit the expected file naming of run_unifold.sh.
> I had to manually fix the structure of the uniprot database.

Thanks for your issue, could you please confirm the correct path of uniprot database as following?
`    --uniprot_database_path=$database_dir/uniprot/uniprot.fasta`

**Comment by dominik-handler** [comment]
That should be correct. Thank you!

**Comment by BaozCWJ** [comment]
> That should be correct. Thank you!

Thx! It's already fixed in the #42
